### PR TITLE
Track outdated nodes per bucket space

### PR DESCRIPTION
**Pull request by toregge**
@geirst, @vekterli: please review

### CODE REVIEW

**Review by geirst — APPROVED** [review]
👍

File: storage/src/vespa/storage/distributor/pendingclusterstate.h [hunk 1]
```diff
@@ -25,6 +25,8 @@ class DistributorBucketSpaceRepo;
  */
 class PendingClusterState : public vespalib::XmlSerializable {
 public:
+    using OutdatedNodes = std::unordered_set<uint16_t>;
+    using OutdatedNodesMap = std::unordered_map<document::BucketSpace, OutdatedNodes, document::BucketSpace::hash>;
```
Comment (vekterli) [inline storage/src/vespa/storage/distributor/pendingclusterstate.h]: This typedef should ideally also be de-duplicated

File: storage/src/vespa/storage/distributor/pending_bucket_space_db_transition.cpp [hunk 2]
```diff
@@ -11,20 +12,42 @@ LOG_SETUP(".pendingbucketspacedbtransition");
...
+using lib::NodeType;
+using lib::NodeState;
+
 PendingBucketSpaceDbTransition::PendingBucketSpaceDbTransition(const PendingClusterState &pendingClusterState,
+                                                               DistributorBucketSpace &distributorBucketSpace,
+                                                               int distributorIndex,
+                                                               bool distributionChanged,
+                                                               const OutdatedNodes &outdatedNodes,
                                                                std::shared_ptr<const ClusterInformation> clusterInfo,
+                                                               const lib::ClusterState &prevClusterState,
```
Comment (vekterli) [inline storage/src/vespa/storage/distributor/pending_bucket_space_db_transition.cpp]: This constructor argument list size is growing outside my comfort zone 😨  We should consider if we can encapsulate this differently during subsequent refactoring.

**Review by vekterli — APPROVED** [review]
LGTM, see comments on typedef duplications

File: storage/src/vespa/storage/distributor/pending_bucket_space_db_transition.cpp [hunk 3]
```diff
@@ -11,20 +12,42 @@ LOG_SETUP(".pendingbucketspacedbtransition");
 
 namespace storage::distributor {
 
+using lib::Node;
+using lib::NodeType;
+using lib::NodeState;
+
 PendingBucketSpaceDbTransition::PendingBucketSpaceDbTransition(const PendingClusterState &pendingClusterState,
+                                                               DistributorBucketSpace &distributorBucketSpace,
+                                                               int distributorIndex,
```
[the remark this answers is not in the document]
Comment (toregge) [reply storage/src/vespa/storage/distributor/pending_bucket_space_db_transition.cpp]: Distributor::getDistributorIndex() returns int. Should that be fixed too ?

**Review by vekterli — APPROVED** [review]
:+1:

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**